plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Purity.hs

```diff
@@ -100,66 +114,80 @@ termEvaluationOrder
   -> EvalOrder name uni fun a
 termEvaluationOrder builtinSemanticsVariant = goTerm
  where
+  goTerm :: Term name uni fun a -> EvalOrder name uni fun a
   goTerm = \case
-    t@(splitApplication -> (Builtin _ann fun, args)) ->
-      foldMap (goTerm . snd) args <> evalOrder
+    t@(splitArgs -> (Builtin _ann fun, args)) ->
+      foldMap goTerm [ a | ApplyTerm a <- args ] <> termEvalOrder parameters args
      where
-      evalOrder =
-        if length args < length (builtinArity @uni @fun (Proxy @uni) builtinSemanticsVariant fun)
-          then -- If it's unsaturated, we definitely don't do any work
-            evalThis (EvalTerm Pure WorkFree t)
-          else -- If it's saturated or oversaturated, we might have an effect here
-            evalThis (EvalTerm MaybeImpure MaybeWork t)
+      parameters :: [Param] =
+        builtinArity @uni @fun (Proxy @uni) builtinSemanticsVariant fun
+
+      termEvalOrder :: [Param] -> [Arg name uni fun a] -> EvalOrder name uni fun a
+      termEvalOrder [] _ =
+        singletonEvalOrder (EvalTerm MaybeImpure MaybeWork t)
+      -- Builtin is not fully saturated with term arguments, thus impure.
```

Comment:
"thus impure" should be "thus pure"

Reply:
My bad: this comment relates to the line below it.
First I had blank lines in between the statements so it was clear that comment relates to the "closest" line, which is the one below. Then I decided to make code more vertically compact and the relation got ambiguous.